src/apps/ipsec/ipsec.lua

```diff
@@ -0,0 +1,155 @@
+module(..., package.seeall)
+local datagram = require("lib.protocol.datagram")
+local ethernet = require("lib.protocol.ethernet")
+local esp = require("lib.protocol.esp")
+local esp_tail = require("lib.protocol.esp_tail")
+local aes_128_gcm = require("apps.ipsec.aes_128_gcm")
+local lib = require("core.lib")
+local ffi = require("ffi")
+
+
+local esp_nh = 50 -- https://tools.ietf.org/html/rfc4303#section-2
+local esp_length = esp:sizeof()
+local esp_tail_length = esp_tail:sizeof()
+
+function esp_v6_new (arg)
+   local conf = arg and config.parse_app_arg(arg) or {}
+   assert(conf.mode == "aes-128-gcm", "Only supports aes-128-gcm.")
+   return { aes_128_gcm = aes_128_gcm:new(conf), seq_no = 0 }
+end
+
+
+local esp_v6_encrypt = {}
+
+function esp_v6_encrypt:new (arg)
+   local o = esp_v6_new(arg)
+   o.pad_buf = ffi.new("uint8_t[?]", o.aes_128_gcm.blocksize-1)
+   o.esp_buf = ffi.new("uint8_t[?]", o.aes_128_gcm.aad_size)
+   -- Fix me https://tools.ietf.org/html/rfc4303#section-3.3.3
+   o.esp = esp:new_from_mem(o.esp_buf, esp_length)
+   o.esp:spi(0x0) -- Fix me, set esp:spi value.
+   o.esp_tail = esp_tail:new({})
+   return setmetatable(o, {__index=esp_v6_encrypt})
+end
+
+-- Return next sequence number.
+function esp_v6_encrypt:next_seq_no ()
+   self.seq_no = self.seq_no + 1
+   return self.seq_no
+end
+
+function esp_v6_encrypt:encrypt (nh, payload, length)
```

Comment:
To talk about efficiency we need a benchmark. Maybe `Source -> Encrypt -> Decrypt -> Sink`? It would be interesting to compare the throughput of the ESP apps with the expectations from Intel's whitepaper and OpenSSL's self-benchmarks.

(In this case I'd say that `packet.append()` already does work that way really. It's `packet.prepend()` that moves data around to make room at the beginning.)